extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java

```diff
@@ -82,6 +82,23 @@ public <T> CommonPanacheQueryImpl<T> project(Class<T> type) {
             throw new PanacheQueryException("Unable to perform a projection on a named query");
         }
 
+        if (query.toLowerCase().startsWith("select new ")) {
+            throw new PanacheQueryException("Unable to perform a projection on a 'select new' query");
+        }
+
+        if (query.toLowerCase().startsWith("select ")) {
+            int endSelect = query.indexOf(" from ");
+            String from = query.substring(endSelect);
+            StringBuilder newQuery = new StringBuilder("SELECT new ")
+                    .append(type.getName())
+                    .append(" (")
+                    .append(query.substring("SELECT ".length(), endSelect))
+                    .append(")")
+                    .append(from);
+
+            return new CommonPanacheQueryImpl<>(this, newQuery.toString(), "select count(*) " + from);
+        }
```

Comment:
I wonder if it'll cover a lot of use case of not.

I saw your test and it seems to working fine for some usage but we already have some request to support query that returns arbitrary types, `SELECT new` will not work for basic types like primitive or String. I wonder if we shouldn't use `jpaQuery.setResultClass()` instead, it will be a more complex change but will be more flexible.

Reply:
What makes you say it won't work for  basic types?
Maybe I'm wrong, but `.setResultClass` doesn't transform the result, it's used to specify what the return type is going to be.

I wanted to write more tests. After spending so much time trying to make the test suite work and fail, I surrendered.

Reply:
You're right, it should be tested.

@DavideD what causes issue to run the Hibernate with Panache integration-test ? It uses and embedded database so it's normally working fine.

Reply:
I've added some test cases for when the select contains constants or other functions. It seems to work as expected.
I think we can merge this.